patches/server/1056-Adding-PredicateChoice-to-Paper-API.patch

```diff
@@ -0,0 +1,212 @@
++        this.values = bukkitChoices.stream().map(CraftItemStack::asNMSCopy).map(ItemValue::new).toArray(Value[]::new);
++    }
++    // Paper end - Adding PredicateChoice
+     public Ingredient(Stream<? extends Ingredient.Value> entries) {
+         this.values = (Ingredient.Value[]) entries.toArray((i) -> {
+             return new Ingredient.Value[i];
+@@ -70,6 +79,11 @@ public final class Ingredient implements Predicate<ItemStack> {
+         } else if (this.isEmpty()) {
+             return itemstack.isEmpty();
+         } else {
++            // Paper start - Adding PredicateChoice
++            if (itemPredicate != null) {
++                return itemPredicate.test(itemstack.getBukkitStack());
++            }
++            // Paper end - Adding PredicateChoice
+             ItemStack[] aitemstack = this.getItems();
+             int i = aitemstack.length;
+ 
+@@ -269,4 +283,10 @@ public final class Ingredient implements Predicate<ItemStack> {
+             return Collections.singleton(this.item);
+         }
+     }
++    // Paper start - Adding PredicateChoice
++    @Nullable
++    public RecipeChoice.PredicateChoice.ItemPredicate getItemPredicate() {
++        return itemPredicate;
++    }
++    // Paper end - Adding PredicateChoice
+ }
+diff --git a/src/main/java/net/minecraft/world/item/crafting/Recipe.java b/src/main/java/net/minecraft/world/item/crafting/Recipe.java
+index 80387cd1bee2bd4c024073cee74222828f9f2c17..6fbbf255fe061943b72324d1484c7b9e885f8535 100644
+--- a/src/main/java/net/minecraft/world/item/crafting/Recipe.java
++++ b/src/main/java/net/minecraft/world/item/crafting/Recipe.java
+@@ -67,7 +67,7 @@ public interface Recipe<C extends Container> {
+     org.bukkit.inventory.Recipe toBukkitRecipe(org.bukkit.NamespacedKey id); // CraftBukkit
+ 
+     // Paper start - improved exact choice recipes
+-    default boolean hasExactIngredients() {
++    default boolean hasSpecialIngredients() {
+         return false;
+     }
+     // Paper end
+diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftRecipe.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftRecipe.java
+index 13d25d118eb4d3ef35a4cdfb9bbde9ed83f6c04b..8b308da84b59c04e0b80d419d4eea602f01abcbf 100644
+--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftRecipe.java
++++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftRecipe.java
+@@ -30,7 +30,13 @@ public interface CraftRecipe extends Recipe {
+         } else if (bukkit instanceof RecipeChoice.ExactChoice) {
+             stack = new Ingredient(((RecipeChoice.ExactChoice) bukkit).getChoices().stream().map((mat) -> new net.minecraft.world.item.crafting.Ingredient.ItemValue(CraftItemStack.asNMSCopy(mat))));
+             stack.exact = true;
+-        } else {
++        }
++        // Paper start - Adding PredicateChoice
++        else if(bukkit instanceof RecipeChoice.PredicateChoice predicateChoice){
++            stack = new Ingredient(predicateChoice.getItemPredicate());
```

Comment:
Just use the regular constructor and pass in the recipe book examples as itemstacks and then assign the predicate field on the next line.